packages/validation/src/MonokleValidator.ts

```diff
@@ -278,16 +281,64 @@ export class MonokleValidator implements Validator {
     }
 
     const run: ValidationRun = {
+      automationDetails: {guid: uniqueId()},
       tool,
       results,
       taxonomies: [NSA_TAXONOMY, CIS_TAXONOMY],
     };
 
-    return {
+    const result: ValidationResponse = {
       $schema: 'https://json.schemastore.org/sarif-2.1.0.json',
       version: '2.1.0',
       runs: [run],
     };
+
+    if (baseline) {
+      this.compareWithBaseline(result, baseline);
+    }
+
+    return result;
+  }
+
+  /**
+   * Compares the validation result with a baseline.
+   *
+   * @remarks comparison uses fingerprints.
+   * @remark enhances the result with baseline guid & state.
+   */
+  private compareWithBaseline(result: ValidationResponse, baseline: ValidationResponse): void {
+    for (const run of result.runs) {
+      const baselineRun = baseline.runs.find(r => r.tool.driver.name === run.tool.driver.name);
+      if (!baselineRun) continue;
+      this.compareRunWithBaseline(run, baselineRun);
+    }
+  }
+
+  private compareRunWithBaseline(run: ValidationRun, baseline: ValidationRun): void {
+    run.baselineGuid = baseline.automationDetails.guid;
+
+    const hashmap = new Map<string, ValidationResult | undefined>();
+    for (const result of baseline.results) {
+      const fingerprint = result.fingerprints?.['monokleHash/v1'];
+      if (!fingerprint) continue;
+      hashmap.set(fingerprint, result);
+    }
+
+    for (const result of run.results) {
+      const fingerprint = result.fingerprints?.['monokleHash/v1'];
+      if (!fingerprint) continue;
+      if (hashmap.has(fingerprint)) {
+        result.baselineState = 'unchanged';
+        hashmap.set(fingerprint, undefined);
+      } else {
+        result.baselineState = 'new';
+      }
+    }
+
+    for (const result of hashmap.values()) {
+      if (!result) continue;
+      run.results.push({...result, baselineState: 'absent'});
+    }
```

Comment:
That's how I understand it from reading the specification (please also take [a peek](https://docs.oasis-open.org/sarif/sarif/v2.1.0/csprd01/sarif-v2.1.0-csprd01.html#_Toc10541100) as it's easy to misunderstand). Practically, it indeed means that there is some burden on the client to properly filter / visualise the problems of type 'absent' or they become false positives.

I doubted a bit to include this to avoid breaking changes, but then I realised that this only happens when the client purposely gives us a baseline as input. Thus it's not a breaking change and a client that provides a baseline should know to do this.

I'll actually add documentation at the place where the client passes it as input to emphasise this.

Reply:
Documented. Please have a look and I'll merge this!

Reply:
Thanks for clarification 👍